pkg/controller/garbagecollector/metaonly/metaonly.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metaonly
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer"
+)
+
+func (obj *MetaOnly) GetObjectKind() unversioned.ObjectKind     { return obj }
+func (obj *MetaOnlyList) GetObjectKind() unversioned.ObjectKind { return obj }
+
+type metaOnlyJSONScheme struct{}
+
+// This function can be extended to mapping different gvk to different MetaOnly,
+// which embedded with different version of ObjectMeta. Currently the system
+// only supports v1.ObjectMeta.
+func gvkToMetaOnly(gvk unversioned.GroupVersionKind) runtime.Object {
+	if strings.HasSuffix(gvk.Kind, "List") {
+		return &MetaOnlyList{}
+	} else {
+		return &MetaOnly{}
+	}
+}
+
+func NewMetaOnlyCodecFactory() serializer.CodecFactory {
+	// populating another scheme from api.Scheme, registering every kind with
+	// MetaOnly (or MetaOnlyList).
+	scheme := runtime.NewScheme()
+	allTypes := api.Scheme.AllKnownTypes()
+	for kind := range allTypes {
+		if kind.Version == runtime.APIVersionInternal {
+			continue
+		}
+		metaOnlyObject := gvkToMetaOnly(kind)
+		scheme.AddKnownTypeWithName(kind, metaOnlyObject)
```

Comment:
No, it will return "kind XXX not registered" error. GC needs to add the 3rd party objects to this scheme as the GC discover them, which it doesn't do right now. 
The discover API includes the `Kind` in the response. The discovery API also needs to indicate the version of the ObjectMeta, otherwise GC can only guess.